SPEC.md

```diff
@@ -0,0 +1,176 @@
+
+For the purposes of this proposal, we define two terms very specifically:
+- _container_ can be considered synonymous with a [Linux _network namespace_][namespaces]. What unit this corresponds to depends on a particular container runtime implementation: for example, in implementations of the [App Container Spec][appc-github] like rkt, each _pod_ runs in a unique network namespace. In [Docker][docker], on the other hand, network namespaces generally exist for each separate Docker container.
+- _network_ refers to a group of entities that are uniquely addressable that can communicate amongst each other. This could be either an individual container (as specified above), a machine, or some other network device (e.g. a router). Containers can be conceptually _added to_ or _removed from_ one or more networks.
+
+[rkt-networking-proposal]: https://docs.google.com/a/coreos.com/document/d/1PUeV68q9muEmkHmRuW10HQ6cHgd4819_67pIxDRVNlM/edit#heading=h.ievko3xsjwxd
+[rkt-networking-design]: 
+https://docs.google.com/a/coreos.com/document/d/1CTAL4gwqRofjxyp4tTkbgHtAwb2YCcP14UEbHNizd8g
+[rkt-github]: https://github.com/coreos/rkt
+[namespaces]: http://man7.org/linux/man-pages/man7/namespaces.7.html 
+[appc-github]: https://github.com/appc/spec
+[docker]: https://docker.com 
+
+## CNI Plugin
+
+### Overview
+
+Each CNI plugin is implemented as an executable that is invoked by the container management system (e.g. rkt or Docker).
+
+A CNI plugin is responsible for inserting a network interface into the container network namespace (e.g. one end of a veth pair) and making any necessary changes on the host (e.g. attaching other end of veth into a bridge). It should then assign the IP to the interface and setup the routes consistent with IP Address Management section by invoking appropriate IPAM plugin.
+
+### Parameters
+
+The operations that the CNI plugin needs to support are:
+
+
+- Add container to network
+  - Parameters:
+    - **Container ID**. This is optional but recommended, and should be unique across an administrative domain. For example, an environment with an IPAM system may require that each container is allocated a unique ID and that each IP allocation can thus be correlated back to a particular container. As another example, in appc implementations this would be the _pod ID_.
+    - **Network namespace path**. This represents the path to the network namespace to be added, i.e. /proc/[pid]/ns/net or a bind-mount/link to it.
+    - **Network configuration**. This is a JSON document describing a network to which a container can be joined. The schema is described below.
+    - **Extra arguments**. This allows granular configuration of CNI plugins on a per-container basis.
+    - **Name of the interface inside the container**. This is the name that should be assigned to the interface created inside the container (network namespace); consequently it must comply with the standard Linux restrictions on interface names.
```

Comment:
@thockin First off, your understanding of the operation is correct.

The reason we wanted to support multiple plugins (or more precisely, joining multiple networks) without writing wrapper code is to make it easy for the user (or admin). I think it will be a common enough operation that it'd be great to do that without custom scripts/code. The big motivator was actually rkt itself: we needed the container to always join the "default" network that would be able to reach the metadata service. It then got expanded to also provide the default route (for Internet) via the host with IP masquerade enabled. The user could then specify additional networks (like container-to-container).

The prescriptive design of 1 network = 1 interface was chosen to:
1) Make it easy to give familiar and predictable ethN interface names. Otherwise the plugin would need to probe to see what name is available.
2) To make status/diagnostics for the container runtime easier. We know that every network results in one interface with one IP address (ok, can be two, IPv4 and IPv6).
3)  Since our abstraction is "joining a network", it seemed like it would be difficult to do so without creating an interface. They only way I could see that it could be accomplished is by re-using an existing interface. However, this would make networks not independent. I can see a case for creating multiple interfaces to join a single network. But that should be rare enough that it can be worked around. For example, go ahead and create the first one with CNI_IFNAME, subsequent ones with random letters and report up only the first one.